src/integrationTests/async.spec.ts

```diff
@@ -23,201 +23,233 @@ beforeEach(() => {
+  describeFn('Using cookies ', () => {
+    beforeEach(() => {
+      useCookie = true;
     });
+    test();
   });
-
-  describe("when initalising with a non-expired identity which requires a refresh", () => {
-    test("it should resolve updated advertising", () => {
-      const originalIdentity = makeIdentity({
-        refresh_from: Date.now() - 100000,
-      });
-      const updatedIdentity = makeIdentity({
-        advertising_token: "updated_advertising_token",
-      });
-      const p = uid2.getAdvertisingTokenAsync();
-      uid2.init({ identity: originalIdentity });
-      xhrMock.responseText = btoa(
-        JSON.stringify({ status: "success", body: updatedIdentity })
-      );
-      xhrMock.onreadystatechange(new Event(""));
-      return expect(p).resolves.toBe(updatedIdentity.advertising_token);
+  describeFn('Using local storage ', () => {
+    beforeEach(() => {
+      useCookie = false;
     });
+    test();
   });
+};
+
+testCookieAndLocalStorage(() => {
+  describe("when getAdvertisingTokenAsync is called before init", () => {
+    describe("when initialising with a valid identity", () => {
+      const identity = makeIdentity();
+      test("it should resolve promise after invoking the callback", () => {
+        const p = uid2.getAdvertisingTokenAsync().then((token: any) => {
+          expect(callback).toHaveBeenCalled();
+          return token;
+        });
+        uid2.init({ callback: callback, identity: identity, useCookie: useCookie });
+        jest.runAllTimers();
+        return expect(p).resolves.toBe(identity.advertising_token);
+      });
+    });
 
-  describe("when auto refresh fails, but identity still valid", () => {
-    test("it should resolve original advertising token", () => {
-      const originalIdentity = makeIdentity({
-        refresh_from: Date.now() - 100000,
+    describe("when initialising with an invalid identity", () => {
+      test("it should reject promise after invoking the callback", () => {
+        const p = uid2.getAdvertisingTokenAsync().catch((e: any) => {
+          expect(callback).toHaveBeenCalled();
+          throw e;
+        });
+        uid2.init({ callback: callback, useCookie: useCookie });
+        return expect(p).rejects.toBeInstanceOf(Error);
       });
-      const p = uid2.getAdvertisingTokenAsync().then((token: any) => {
-        expect(callback).toHaveBeenCalled();
-        return token;
+    });
+
+    describe("when initialising with a non-expired identity which requires a refresh", () => {
+      test("it should resolve updated advertising", () => {
+        const originalIdentity = makeIdentity({
+          refresh_from: Date.now() - 100000,
+        });
+        const updatedIdentity = makeIdentity({
+          advertising_token: "updated_advertising_token",
+        });
+        const p = uid2.getAdvertisingTokenAsync();
+        uid2.init({ identity: originalIdentity, useCookie: useCookie });
+        xhrMock.responseText = btoa(
```

Comment:
nit:  maybe we could wrap xhrMock.responseText and xhrMock.onreadystatechange into a `mockResponse(response)` function?